aten/src/ATen/cpu/vec/functional_base.h

```diff
@@ -76,6 +76,34 @@ struct VecReduceAllSIMD<float, Op> {
   }
 };
 #endif // defined(CPU_CAPABILITY_AVX512)
+#if defined(CPU_CAPABILITY_NEON)
+template <typename Op>
+struct VecReduceAllSIMD<float, Op> {
+  static inline float apply(const Op& vec_fun, const Vectorized<float>& acc_vec) {
+    using Vec = Vectorized<float>;
+    Vec v = acc_vec;
+
+    // 128-bit shuffle: [a1, a2, a3, a4, a5, a6, a7, a8] -> [a5, a6, a7, a8, a1, a2, a3, a4]
+    Vec v1 = {v.get_high(), v.get_low()};
+    // [a1+a5, a2+a6, a3+a7, a4+a8, -, -, -, -] ('+' stands for the reduction function. Note that the last 4 elements are not required)
+    v = vec_fun(v, v1); 
+
+    // 64-bit shuffle: [a1+a5, a2+a6, a3+a7, a4+a8, -, -, -, -] -> [a3+a7, a4+a8, a1+a5, a2+a6, -, -, -, -]
+    float32x4_t v1_1 = vextq_f32(v.get_low(), v.get_low(), 2);
+    v1 = {v1_1, v1_1};
+    // [a1+a3+a5+a7, a2+a4+a6+a8, a1+a3+a5+a7, a2+a4+a6+a8, -, -, -, -]
+    v = vec_fun(v, v1); 
+
+    // 32-bit shuffle: [a1+a3+a5+a7, a2+a4+a6+a8, a1+a3+a5+a7, a2+a4+a6+a8, -, -, -, -] -> [a2+a4+a6+a8, a1+a3+a5+a7, a2+a4+a6+a8, a1+a3+a5+a7, -, -, -, -]
+    v1_1 = vrev64q_f32(v.get_low());
+    v1 = {v1_1, v1_1};
+    // [a1+a2+a3+a4+a5+a6+a7+a8, a1+a2+a3+a4+a5+a6+a7+a8, a1+a2+a3+a4+a5+a6+a7+a8, a1+a2+a3+a4+a5+a6+a7+a8, -, -, -, -]
+    v = vec_fun(v, v1);
+
+    return v.get_low()[0];
+  }
+};
+#endif // defined(CPU_CAPABILITY_NEON)
```

Comment:
Reference AVX-2 implementation: [here](https://github.com/pytorch/pytorch/blob/ced5c89b6fbe827a538b7ada96b2f9a5989871c7/aten/src/ATen/cpu/vec/functional_base.h#L38-L56)

Reply:
Do we have UT to cover this?

Reply:
@jgong5 , I need some help in writing the UT for this.

One option is to test if the accuracy of reduce_all or softmax is comparable in eager and compiled mode. Let me know if this is acceptable.

Other option is to extend Intel's AVX `vec_reduce_all` UT. From what I understood from this [comment](https://github.com/pytorch/pytorch/pull/73953#issuecomment-1121820121), the UT for Intel's `vec_reduce_all` is accomplished by `vec_test_all_types_AVX2`, `vec_test_all_types_AVX512` and `vec_test_all_types_DEFAULT`. From the build logs, I see that these test executables are built by cmake. My guess is that if I need to extend the existing UT structure and generate a similar executable for NEON, I would need to modify certain **critical build files** such as this and more cc @malfet https://github.com/pytorch/pytorch/blob/795885d9476a4978c476b221720332b73d994d7d/cmake/Codegen.cmake#L315-L350 
Although my current code passes the [FunctionalTestsReducedFloat test](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/test/vec_test_all_types.cpp#L1429), it goes through the default path. Could you share some more details on how I should go about adding a UT for this reduction function? 

Thanks!

Reply:
@jgong5 , I need some help in writing the UT for this.

One option is to test if the accuracy of `reduce_all` or `softmax` is comparable in eager and compiled mode. Let me know if this is acceptable.

Other option is to extend Intel's AVX `vec_reduce_all` UT. From what I understood from this [comment](https://github.com/pytorch/pytorch/pull/73953#issuecomment-1121820121), the UT for Intel's `vec_reduce_all` is accomplished by `vec_test_all_types_AVX2`, `vec_test_all_types_AVX512` and `vec_test_all_types_DEFAULT`. From the build logs, I see that these test executables are built by cmake. My guess is that if I need to extend the existing UT structure and generate a similar executable for NEON, I would need to modify certain **critical build files** such as this and more. https://github.com/pytorch/pytorch/blob/795885d9476a4978c476b221720332b73d994d7d/cmake/Codegen.cmake#L315-L350
Although my current code passes the [FunctionalTestsReducedFloat test](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/test/vec_test_all_types.cpp#L1429), it goes through the default path. Could you share some more details on how I should go about adding a UT for this reduction function?

Thanks!

Reply:
Additionally, do the CI UTs run on x86 machines? Could you throw some light on how NEON tests are executed? cc @malfet